src/lib/NonLinearDutchOrderLib.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {DutchOrderLib} from "./DutchOrderLib.sol";
+import {OrderInfo} from "../base/ReactorStructs.sol";
+import {OrderInfoLib} from "./OrderInfoLib.sol";
+import {ERC20} from "solmate/src/tokens/ERC20.sol";
+
+struct CosignerData {
+    // The block at which the input or outputs start decaying
+    uint256 decayStartBlock;
+    // The address who has exclusive rights to the order until decayStartBlock
+    address exclusiveFiller;
+    // The amount in bps that a non-exclusive filler needs to improve the outputs by to be able to fill the order
+    uint256 exclusivityOverrideBps;
+    // The tokens that the swapper will provide when settling the order
+    uint256 inputAmount;
+    // The tokens that must be received to satisfy the order
+    uint256[] outputAmounts;
+}
+
+struct NonLinearDutchOrder {
+    // generic order information
+    OrderInfo info;
+    // The address which must cosign the full order
+    address cosigner;
+    // The tokens that the swapper will provide when settling the order
+    NonLinearDutchInput baseInput;
+    // The tokens that must be received to satisfy the order
+    NonLinearDutchOutput[] baseOutputs;
+    // signed over by the cosigner
+    CosignerData cosignerData;
```

Comment:
since these are purely relative, the cosigner now has full power to set the start amount to whatever they want? some thoughts here:
- this is a lot more power than cosigner has had in the past. Maybe we want to consider setting a min/max value?
- potentially cosigned start/end amounts could be so high/low to cause under/overflow after adjustment - probably want to make sure we cover that case

Reply:
The cosigner can only update the startAmount (not the relative amounts) and only if it's better than what the swapper signed ([see here](https://github.com/Uniswap/UniswapX/blob/295575ff548bc743e5a982e1a0339b268f350920/src/reactors/NonLinearDutchOrderReactor.sol#L95)). The curve and base amounts are set by the swapper. Because of this, I don't think the cosigner has more trust than it did in the past. LMK if I'm missing something.

If the cosigner is malicious and attempts to cause overflow, the tx will revert ([as it's currently implemented](https://github.com/Uniswap/UniswapX/blob/295575ff548bc743e5a982e1a0339b268f350920/test/lib/NonLinearDutchDecayLib.t.sol#L250)). Underflow isn't possible since we're subtracting an `int256` from a `uint256`. In the case of overflow we could either:
1. Do nothing and have the swapper deal with the failed tx (eg. use another cosigner).
2. Detect the overflow in the reactor and fallback to using the user provided input. This will be more gas intensive since we'd need to loop through the curve and check each point. 

Given this could only occur with a malicious cosigner and the worst case is that the tx reverts, I think it's safe to just let the tx revert.